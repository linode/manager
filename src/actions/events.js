export const RESET_EVENTS_POLL = '@@events/RESET_EVENTS_POLL';

export function resetEventsPoll() {
  return { type: RESET_EVENTS_POLL };
}
