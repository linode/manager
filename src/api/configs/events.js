import {
  genConfig, ReducerGenerator, genActions, ONE, MANY,
} from '~/api/gen';

export const config = genConfig({
  plural: 'events',
  singular: 'event',
  endpoint: id => `/account/events/${id}`,
  supports: [ONE, MANY],
});

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
