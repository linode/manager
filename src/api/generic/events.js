import {
  ReducerGenerator, genActions, ONE, MANY,
} from '~/api/internal.ts';

export const config = {
  name: 'events',
  primaryKey: 'id',
  endpoint: id => `/account/events/${id}`,
  supports: [ONE, MANY],

  // sort desc by created
  sortFn: function (ids, state) {
    return ids.sort(function (a, b) {
      const aDate = new Date(state[a].created);
      const bDate = new Date(state[b].created);

      if (aDate > bDate) {
        return -1;
      }

      if (aDate < bDate) {
        return 1;
      }

      return 0;
    });
  },
};

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
