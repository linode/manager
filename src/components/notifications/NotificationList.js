import React, { PropTypes, Component } from 'react';

import NotificationListItem from './NotificationListItem';


export default class NotificationList extends Component {

  constructor(props) {
    super(props);

    this.onClickShowMore = this.onClickShowMore.bind(this);

    this.state = { totalResults: props.events.totalResults || 0 };
  }

  componentWillUpdate(nextProps) {
    this.updateTotalResults(nextProps.events);
  }

  onClickShowMore(e) {
    e.preventDefault();
    this.props.onClickShowMore(e);
  }

  updateTotalResults(events) {
    if (events.totalResults > this.state.totalResults) {
      this.setState({ totalResults: events.totalResults });
    }
  }

  render() {
    const { events, loading = false, open, onClickItem } = this.props;
    const eventMap = events.events;
    const showMore = events.ids.length < this.state.totalResults;
    let message;

    if (loading) {
      message = (<span className="LoadingMessage text-muted">Loading...</span>);
    } else if (showMore) {
      message = (
        <span className="ShowMoreLink btn btn-link">Show more</span>
      );
    } else {
      message = 'No more notifications.';
    }

    return (
      <div className={`NotificationList ${open ? 'NotificationList--open' : ''}`}>
        <div className="NotificationList-body">
          <div>
            {events.ids.map((id, index) =>
              <NotificationListItem
                key={index}
                onClick={onClickItem}
                event={eventMap[id]}
              />)}
            <div
              className="NotificationList-listItem NotificationList-end text-sm-center"
              onClick={showMore ? this.onClickShowMore : null}
            >
              {message}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

NotificationList.propTypes = {
  events: PropTypes.object.isRequired,
  loading: PropTypes.bool,
  open: PropTypes.bool.isRequired,
  onClickItem: PropTypes.func,
  onClickShowMore: PropTypes.func,
};

NotificationList.defaultProps = {
  open: false,
};
