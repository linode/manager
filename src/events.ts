import * as moment from 'moment';
import { assoc, compose, either, ifElse, isEmpty, isNil, lensPath, map, not, over, path, view, when } from 'ramda';
import { Subject } from 'rxjs/Subject';

import { DISABLE_EVENT_THROTTLE } from 'src/constants';
import { getEvents } from 'src/services/account';
import { dateFormat } from 'src/time';
import isPast from 'src/utilities/isPast';

const createInitialDatestamp = () => {
  return moment('1970-01-01 00:00:00.000Z').utc().format(dateFormat);
}

export const events$ = new Subject<Linode.Event>();

let filterDatestamp = createInitialDatestamp();
const pollIDs: { [key: string]: boolean } = {};

const initialPollInterval = 2000;
export let eventRequestDeadline = Date.now();
export let currentPollIntervalMultiplier = 1;

export const resetEventsPolling = () => {
  eventRequestDeadline = Date.now() + initialPollInterval;
  currentPollIntervalMultiplier = 1;
}
export const init = () => {
  filterDatestamp = createInitialDatestamp();
  resetEventsPolling();
};

export const generateInFilter = (keyName: string, arr: any[]) => {
  return {
    '+or': arr.map(el => ({ [keyName]: el })),
  };
}

export const generatePollingFilter = (datestamp: string, pollIDsAsArr: string[]) => {
  return pollIDsAsArr.length ?
    {
      '+or': [
        { created: { '+gt': datestamp } },
        generateInFilter('id', pollIDsAsArr),
      ],
    }
    : {
      created: { '+gt': datestamp },
    };
}

const theBeginningOfTime = moment.utc('1970-01-01 00:00:00.000Z').format();
const isPasttheBeginningOfTime = isPast(theBeginningOfTime);

/**
 * If the X-Filter is set we parse it and check for and compare the created.+gt value to
 * "the beginning of time". If the value is greater, do nothing, otherwise update the events to have
 * an _initial prop of true.
 */

export const setInitialEvents = when(
  compose(not, isNil, view(lensPath(['config', 'headers', 'X-Filter']))),

  (response) => {
    try {
      return when(
        compose(
          ifElse(
            isNil,
            () => false,
            compose(not, isPasttheBeginningOfTime),
          ),
          either(path(['created', '+gt']) as any, path(['+or', 0, 'created', '+gt']) as any),
          when(compose(not, isEmpty), v => JSON.parse(v)),
          path(['config', 'headers', 'X-Filter']),
        ),
        over(lensPath(['data', 'data']), map(assoc('_initial', true))),
      )(response);

    } catch (error) { return response; }
  },
);

export const requestEvents = () => {
  getEvents(
    { page_size: 25 },
    generatePollingFilter(filterDatestamp, Object.keys(pollIDs)),
  )
    .then(setInitialEvents)
    .then(response => response.data.data)
    .then((data) => {
      /*
        * Events come back in reverse chronological order, so we update our
        * datestamp with the latest Event that we've seen. We need to perform
        * a date comparison here because we also might get back some old events
        * from IDs that we're polling for.
        */

      if (data[0]) {
        const newDatestamp = moment(data[0].created);
        const currentDatestamp = moment(filterDatestamp);
        if (newDatestamp > currentDatestamp) {
          filterDatestamp = newDatestamp.format(dateFormat);
        }
      }

      data.reverse().map((linodeEvent: Linode.Event, idx: number, events: Linode.Event[]) => {
        // if an Event completes it is removed from pollIDs
        if (linodeEvent.percent_complete === 100
          && pollIDs[linodeEvent.id]) {
          delete pollIDs[linodeEvent.id];
        }

        // we poll for Event IDs that have not yet been completed
        if (
          linodeEvent.percent_complete !== null
          && linodeEvent.percent_complete < 100

          /** If a Linode is deleted the /events end-points sends updated regarding shutting down
           * and eventuallly deletion. Subscribers of this stream want active Linodes only, and
           * if provided a "deleted" Linode ID will result in 404s until the events stop.
           */
          && !isBeingDeleted(events, linodeEvent.id)
        ) {
          // when we have an "incomplete event" poll at the initial polling rate
          resetEventsPolling();
          pollIDs[linodeEvent.id] = true;
        }

        events$.next(linodeEvent);
      });
    });
}

setInterval(
  () => {
    if (Date.now() > eventRequestDeadline) {
      requestEvents();

      if (DISABLE_EVENT_THROTTLE) {
        /* If throttling is disabled, don't use or update the multiplier */
        const mocksPollingInterval = 500;

        eventRequestDeadline = Date.now() + mocksPollingInterval;
      } else {
        eventRequestDeadline = Date.now()
          + initialPollInterval * currentPollIntervalMultiplier;
        /* double the polling interval with each poll up to 16x */
        currentPollIntervalMultiplier = Math.min(currentPollIntervalMultiplier * 2, 16);
      }
    }
  },
  /* the following is the Nyquist rate for the minimum polling interval */
  (initialPollInterval / 2 - 1),
);

const isBeingDeleted = (events: Linode.Event[], id: number): boolean =>
  events.filter(event => event.id === id && event.action === 'linode_delete').length > 0;
