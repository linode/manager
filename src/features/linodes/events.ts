import * as moment from 'moment';

export const newLinodeEvents = (mountTime: moment.Moment) =>
(linodeEvent : Linode.Event): boolean => {
  const actionWhitelist = [
    'linode_boot',
    'linode_reboot',
    'linode_shutdown',
    'backups_enable',
    'backups_cancel',
    'backups_restore',
    'linode_snapshot',
    'linode_rebuild',
    'linode_resize',
    'disk_resize',
    'linode_migrate',
  ];

  const statusWhitelist = [
    'started',
    'finished',
    'scheduled',
    'failed',
    'notification',
  ];

  const isLinodeEvent = linodeEvent.entity !== null && linodeEvent.entity.type === 'linode';

  const result = isLinodeEvent
    && statusWhitelist.includes(linodeEvent.status)
    && actionWhitelist.includes(linodeEvent.action)
    && (!linodeEvent._initial);

  return result;
};
