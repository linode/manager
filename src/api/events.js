import { fetch } from '~/fetch';
import { actions } from './configs/events';

export function eventAction(action) {
  return (eventId) => async (dispatch, getState) => {
    const state = getState();

    const event = state.api.events.events[eventId];
    await dispatch(actions.one({ ...event, [action]: true }, eventId));

    if (action === 'seen') {
      // Mark all events seen. Using many to avoid a dispatch per event.
      const page = {
        events: [],
        total_pages: state.api.events.totalPages,
        total_results: state.api.events.totalResults,
      };
      Object.values(state.api.events.events).forEach(e => {
        if (!e.seen) {
          page.events.push({ ...e, seen: true});
        }
      });
      dispatch(actions.many(page));
    }

    const { token } = state.authentication;
    fetch(token, `/account/events/${eventId}/${action}`, { method: 'POST' });
  };
}

export const eventRead = eventAction('read');
export const eventSeen = eventAction('seen');
