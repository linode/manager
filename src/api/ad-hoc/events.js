import { actions } from '../generic/events';
import { fetch } from '../fetch';


export function eventAction(action) {
  return (eventId) => async (dispatch, getState) => {
    const state = getState();

    if (action === 'seen') {
      // Mark all events seen. Using many to avoid a dispatch per event.
      const page = {
        pages: state.api.events.totalPages,
        results: state.api.events.totalResults,
        events: Object.values(state.api.events.events).reduce(
          (unseen, e) => e.seen ? unseen : [...unseen, { ...e, seen: true }], []) || [],
      };
      dispatch(actions.many(page));
    } else {
      const event = state.api.events.events[eventId];
      await dispatch(actions.one({ ...event, [action]: true }, eventId));
    }

    return dispatch(fetch.post(`/account/events/${eventId}/${action}`));
  };
}

export const eventRead = eventAction('read');
export const eventSeen = eventAction('seen');
