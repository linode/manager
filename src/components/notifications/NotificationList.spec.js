import React from 'react';
import { mount } from 'enzyme';
import sinon from 'sinon';
import { StaticRouter } from 'react-router-dom';

import { NotificationList } from './NotificationList';

import { api, fakeAPI } from '~/data';
import { events } from '~/data/events';


// NOTE: item click handling untested because of <Link> usage
describe('components/notifications/NotificationList', () => {
  const sandbox = sinon.sandbox.create();
  let notificationList;

  afterEach(() => {
    sandbox.restore();
  });

  it('implements default state', () => {
    const wrapper = mount(
      <StaticRouter>
        <NotificationList events={api.events} />
      </StaticRouter>
    );

    const notificationList = wrapper.find('NotificationList');

    expect(notificationList.instance().state).toBeDefined();
    expect(notificationList.instance().state.totalResults).toBe(4);
  });

  it('updates total results based on events results', () => {
    let eventsAPI = fakeAPI([[events, 'event', 'events', 200]]);
    const wrapper = mount(
      <StaticRouter>
        <NotificationList events={eventsAPI.events} />
      </StaticRouter>
    );

    const notificationList = wrapper.find('NotificationList');
    expect(notificationList.instance().state.totalResults).toBe(200);

    eventsAPI = fakeAPI([[events, 'event', 'events', 250]]);
    notificationList.instance().updateTotalResults(eventsAPI.events);

    expect(notificationList.instance().state.totalResults).toBe(250);
  });

  it('updates total results but also retains the max number of results seen', () => {
    let eventsAPI = fakeAPI([[events, 'event', 'events', 200]]);
    const wrapper = mount(
      <StaticRouter>
        <NotificationList
          events={eventsAPI.events}
        />
      </StaticRouter>
    );
    const notificationList = wrapper.find('NotificationList');

    expect(notificationList.instance().state.totalResults).toBe(200);

    eventsAPI = fakeAPI([[events, 'event', 'events', 100]]); // less total
    notificationList.instance().updateTotalResults(eventsAPI.events);

    expect(notificationList.instance().state.totalResults).toBe(200);
  });

  it('renders events', () => {
    const notifications = mount(
      <StaticRouter>
        <NotificationList events={api.events} />
      </StaticRouter>
    );
    const count = Object.values(api.events.events).filter(e => !e.seen).length;

    expect(notifications.find('NotificationListItem').length).toBe(count);

    // This event was updated last.
    expect(notifications.find('NotificationListItem').at(0).props().event.id).toBe(385);
  });


  it('renders a show more link when more results are available', () => {
    const eventsAPI = fakeAPI([
      [events, 'event', 'events', 200],
    ]);
    notificationList = mount(
      <StaticRouter>
        <NotificationList
          events={eventsAPI.events}
        />
      </StaticRouter>
    );

    expect(notificationList.find('.ShowMoreLink').length).toBe(1);
  });

  it('supports show more click handling', () => {
    const clickShowMoreHandlerSpy = sandbox.spy();
    const eventsAPI = fakeAPI([
      [events, 'event', 'events', 200],
    ]);
    notificationList = mount(
      <StaticRouter>
        <NotificationList
          events={eventsAPI.events}
          onClickShowMore={clickShowMoreHandlerSpy}
        />
      </StaticRouter>
    );

    notificationList.find('.ShowMoreLink').first().simulate('click');

    expect(clickShowMoreHandlerSpy.callCount).toBe(1);
  });

  it('should show a loading message when flag is set', () => {
    const notifications = mount(
      <StaticRouter>
        <NotificationList
          events={api.events}
          loading
        />
      </StaticRouter>
    );

    expect(notifications.find('.LoadingMessage').length).toBe(1);
  });
});
