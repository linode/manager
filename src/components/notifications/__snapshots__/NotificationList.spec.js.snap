// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/notifications/NotificationList should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StaticRouter
    basename=""
    location="/"
  >
    <NotificationList
      events={
        Object {
          "events": Object {
            "385": Object {
              "action": "linode_reboot",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 385,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": false,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:03:40",
            },
            "386": Object {
              "action": "linode_reboot",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 386,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": true,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:12:40",
            },
            "387": Object {
              "action": "linode_shutdown",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 385,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": false,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:03:40",
            },
            "388": Object {
              "action": "linode_boot",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 385,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": false,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:03:40",
            },
          },
          "ids": Array [
            385,
            386,
            385,
            385,
          ],
          "pagesFetched": Array [
            1,
          ],
          "plural": "events",
          "singular": "event",
          "totalPages": 1,
          "totalResults": 4,
        }
      }
      open={false}
    />
  </StaticRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <NotificationList
        events={
          Object {
            "events": Object {
              "385": Object {
                "action": "linode_reboot",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 385,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": false,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:03:40",
              },
              "386": Object {
                "action": "linode_reboot",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 386,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": true,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:12:40",
              },
              "387": Object {
                "action": "linode_shutdown",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 385,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": false,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:03:40",
              },
              "388": Object {
                "action": "linode_boot",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 385,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": false,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:03:40",
              },
            },
            "ids": Array [
              385,
              386,
              385,
              385,
            ],
            "pagesFetched": Array [
              1,
            ],
            "plural": "events",
            "singular": "event",
            "totalPages": 1,
            "totalResults": 4,
          }
        }
        open={false}
      />,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "events": Object {
          "events": Object {
            "385": Object {
              "action": "linode_reboot",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 385,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": false,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:03:40",
            },
            "386": Object {
              "action": "linode_reboot",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 386,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": true,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:12:40",
            },
            "387": Object {
              "action": "linode_shutdown",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 385,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": false,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:03:40",
            },
            "388": Object {
              "action": "linode_boot",
              "created": "2016-11-21T20:03:40",
              "entity": Object {
                "label": "linode-www2",
              },
              "finished": null,
              "id": 385,
              "label_message": "",
              "linode_id": 1237,
              "nodebalancer_id": null,
              "percent_complete": 100,
              "read": false,
              "seen": false,
              "stackscript_id": null,
              "status": "finished",
              "updated": "2016-11-21T20:03:40",
            },
          },
          "ids": Array [
            385,
            386,
            385,
            385,
          ],
          "pagesFetched": Array [
            1,
          ],
          "plural": "events",
          "singular": "event",
          "totalPages": 1,
          "totalResults": 4,
        },
        "open": false,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <NotificationList
          events={
            Object {
              "events": Object {
                "385": Object {
                  "action": "linode_reboot",
                  "created": "2016-11-21T20:03:40",
                  "entity": Object {
                    "label": "linode-www2",
                  },
                  "finished": null,
                  "id": 385,
                  "label_message": "",
                  "linode_id": 1237,
                  "nodebalancer_id": null,
                  "percent_complete": 100,
                  "read": false,
                  "seen": false,
                  "stackscript_id": null,
                  "status": "finished",
                  "updated": "2016-11-21T20:03:40",
                },
                "386": Object {
                  "action": "linode_reboot",
                  "created": "2016-11-21T20:03:40",
                  "entity": Object {
                    "label": "linode-www2",
                  },
                  "finished": null,
                  "id": 386,
                  "label_message": "",
                  "linode_id": 1237,
                  "nodebalancer_id": null,
                  "percent_complete": 100,
                  "read": true,
                  "seen": false,
                  "stackscript_id": null,
                  "status": "finished",
                  "updated": "2016-11-21T20:12:40",
                },
                "387": Object {
                  "action": "linode_shutdown",
                  "created": "2016-11-21T20:03:40",
                  "entity": Object {
                    "label": "linode-www2",
                  },
                  "finished": null,
                  "id": 385,
                  "label_message": "",
                  "linode_id": 1237,
                  "nodebalancer_id": null,
                  "percent_complete": 100,
                  "read": false,
                  "seen": false,
                  "stackscript_id": null,
                  "status": "finished",
                  "updated": "2016-11-21T20:03:40",
                },
                "388": Object {
                  "action": "linode_boot",
                  "created": "2016-11-21T20:03:40",
                  "entity": Object {
                    "label": "linode-www2",
                  },
                  "finished": null,
                  "id": 385,
                  "label_message": "",
                  "linode_id": 1237,
                  "nodebalancer_id": null,
                  "percent_complete": 100,
                  "read": false,
                  "seen": false,
                  "stackscript_id": null,
                  "status": "finished",
                  "updated": "2016-11-21T20:03:40",
                },
              },
              "ids": Array [
                385,
                386,
                385,
                385,
              ],
              "pagesFetched": Array [
                1,
              ],
              "plural": "events",
              "singular": "event",
              "totalPages": 1,
              "totalResults": 4,
            }
          }
          open={false}
        />,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "events": Object {
            "events": Object {
              "385": Object {
                "action": "linode_reboot",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 385,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": false,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:03:40",
              },
              "386": Object {
                "action": "linode_reboot",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 386,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": true,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:12:40",
              },
              "387": Object {
                "action": "linode_shutdown",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 385,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": false,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:03:40",
              },
              "388": Object {
                "action": "linode_boot",
                "created": "2016-11-21T20:03:40",
                "entity": Object {
                  "label": "linode-www2",
                },
                "finished": null,
                "id": 385,
                "label_message": "",
                "linode_id": 1237,
                "nodebalancer_id": null,
                "percent_complete": 100,
                "read": false,
                "seen": false,
                "stackscript_id": null,
                "status": "finished",
                "updated": "2016-11-21T20:03:40",
              },
            },
            "ids": Array [
              385,
              386,
              385,
              385,
            ],
            "pagesFetched": Array [
              1,
            ],
            "plural": "events",
            "singular": "event",
            "totalPages": 1,
            "totalResults": 4,
          },
          "open": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
